src/main/java/org/reaktivity/nukleus/tcp/internal/stream/ReadStream.java

```diff
@@ -106,8 +107,20 @@ int handleStream(
             key.cancel(OP_READ);
             if (!closed)
             {
-                closed = true;
-                connectionDone.run();
+                try
+                {
+                    closed = true;
+                    connectionDone.run();
+                    channel.shutdownInput();
+                    if (channel.socket().isOutputShutdown())
+                    {
+                        CloseHelper.quietClose(channel);
+                    }
+                }
+                catch (IOException e)
+                {
+                    LangUtil.rethrowUnchecked(e);
+                }
```

Comment:
Hoist code into methods as shown to support the following snippet replacing above.
```
closed = true;
connectionDone.run();
shutdownInput();
closeIfOutputShutdown();
```
```
private void shutdownInput()
{
    try
    {
        channel.shutdownInput();
    }
    catch (IOException ex)
    {
        LangUtil.rethrowUnchecked(ex);
    }
}
```
```
private void closeIfOutputShutdown()
{
    if (channel.socket().isOutputShutdown()
    {
        CloseHelper.quietClose(channel);
    }
}
```